aptos-move/aptos-vm/src/verifier/event_validation.rs

```diff
@@ -0,0 +1,149 @@
+                    if module_addr
+                        != &AccountAddress::from_hex_literal("0x1").expect("valid address")
+                        || module_name.as_str() != EVENT_MODULE_NAME
+                        || func_name.as_str() != EVENT_EMIT_FUNCTION_NAME
+                    {
+                        continue;
+                    }
+                    let param = module
+                        .signature_at(func_instantiation.type_parameters)
+                        .0
+                        .first()
+                        .ok_or_else(|| {
+                            metadata_validation_error(
+                                "Missing parameter for 0x1::event::emit function",
+                            )
+                        })?;
+                    match param {
+                        StructInstantiation(index, _) | Struct(index) => {
+                            let struct_handle = &module.struct_handle_at(*index);
+                            let struct_name = module.identifier_at(struct_handle.name);
+                            if struct_handle.module != module.self_handle_idx() {
+                                metadata_validation_err(format!("{} passed to 0x1::event::emit function is not defined in the same module", struct_name).as_str())
+                            } else if !event_structs.contains(struct_name.as_str()) {
+                                metadata_validation_err(format!("#[event] attribute is missing for {} passed to 0x1::event::emit function", struct_name).as_str())
+                            } else {
+                                Ok(())
+                            }
+                        },
+                        _ => metadata_validation_err(
+                            "Reference to non-struct parameter for 0x1::event::emit function",
+                        ),
+                    }?;
+                }
+            }
+        }
+    }
+    Ok(())
+}
+
+/// Given a module id extract all event metadata
+pub(crate) fn extract_event_metadata_from_module(
+    session: &mut SessionExt,
+    module_id: &ModuleId,
+) -> VMResult<HashSet<String>> {
+    let metadata = session.load_module(module_id).map(|module| {
+        CompiledModule::deserialize(&module)
+            .map(|module| aptos_framework::get_metadata_from_compiled_module(&module))
+    });
+
+    if let Ok(Ok(Some(metadata))) = metadata {
+        extract_event_metadata(&metadata)
+    } else {
+        Ok(HashSet::new())
+    }
+}
+
+/// Given a module id extract all event metadata
+pub(crate) fn extract_event_metadata(
+    metadata: &RuntimeModuleMetadataV1,
+) -> VMResult<HashSet<String>> {
+    let mut event_structs = HashSet::new();
+    for (struct_, attrs) in &metadata.struct_attributes {
+        for attr in attrs {
+            if attr.is_event() && !event_structs.insert(struct_.clone()) {
+                metadata_validation_err("Found duplicate event attribute")?;
+            }
```

Comment:
again we have this logic where we validate the logic on the metadata struct attributes?

Reply:
do we have that? I don't think so...